Add the ability to recursively download folders from the browse interface

KyokoMiki opened the pull request:
#807 

![a6d57e3914d425108532f63769c02f3e](https://github.com/slskd/slskd/assets/161209740/859bab89-b379-4799-b1f8-5d9f582042f6)

![45cdf41cb4ab9d779c61f418ab2c21bb](https://github.com/slskd/slskd/assets/161209740/4c285d9f-23ad-477d-87b5-3e45d6979276)

![278c6defcd355986930005338be85c9b](https://github.com/slskd/slskd/assets/161209740/1ede96d7-3cb3-49f5-b16d-06d951482e61)

Previously, the entire backend download directory was hard-encoded and could not be manually specified. Therefore, I modified api endpoint:

```
namespace slskd.Transfers.API
{
    public class QueueDownloadRequest
    {
        public string Filename { get; set; }
        public long Size { get; set; }

        public string Path { get; set; }
    }
}
```

I added the ability to specify the download directory (relative path). However, because I modified slskd.Transfers.Transfer, it **caused the transfers.db database to no longer be compatible with the previous version**. The "Path" column needs to be manually added to the transfers.db database.

It should also be possible to use a method similar to subnet masks, adding a small segment after the filename to indicate where the relative path starts. However, I believe manually specifying the download file name is still quite important. In the future, there may be other issues that require this functionality. Therefore, I wanted to submit this first and ask for the developer's opinion.

KyokoMiki commented:
> Thanks for the contribution! I took a look at the changes in the UI and I think they look great!
> 
> I glanced at the backend changes and I'll need to do a more thorough review; path normalization has been contentious at several different points during development and I want to make sure this isn't going to introduce any regressions. I didn't expect that there would need to be a backend change for the application to support this.

Thank you very much!

This issue is because the backend directly specifies the last two parts of the remote directory as the relative path when handling downloads (a/b/c/x.flac -> c/x.flac). Therefore, apart from modifying the API to specify downloading to b/c/x.flac and adding the ability to override paths in the backend, I couldn't come up with a better solution. My personal ability is limited, so there might be better approaches.

lazee486 commented:
just discovered soulseek, and the webui is awesome, hope this feature comes soon :) serious thanks!

ricardopeters-dev commented:
Hi! Thanks for the work on this — just wanted to explain an issue I’m running into with multi-disc albums.

Since it's not yet possible to download folders recursively, I have to manually download each subfolder (like CD1, CD2). The problem is they all get saved into the same top-level folder — for example, /downloads/CD1 — so when I grab another album with the same disc structure, the files end up mixed together or overwritten.

For example, downloading The Beatles – White Album and Smashing Pumpkins – Mellon Collie and the Infinite Sadness results in both albums' CD1 and CD2 tracks ending up in the same place.

It’d be great if full folder structures could be preserved, or if recursive downloading was supported to avoid this kind of conflict.

Appreciate all the progress so far!

KyokoMiki commented:
@jpdillingham I have refactored the recursive download functionality. The implementation in this version is much simpler than before and will no longer produce breaking changes like it used to.

KyokoMiki committed:
Add the ability to recursively download folders from the browse interface

jpdillingham commented:
The backend changes aren't necessary, this can make a single request and include all of the files with their full path, the same as what's done when selecting files and clicking the download button currently.

Where the files go on disk is a separate issue that I'm still thinking about a solution for.

zoeleu commented:
any reason why this hasn't been merged yet?

KyokoMiki committed:
Add recursive folder download from browse and search interfaces

Implement folder download functionality that allows users to download entire directories
with all subdirectories and files from both the browse interface and search results.

- Add download folder button to FileList component with loading state
- Implement recursive directory traversal in Browse component
- Add batch download endpoint integration for multiple files
- Support fallback to search results if directory contents fetch fails
- Add visual feedback with spinner icon during download
- Include destination path mapping to preserve folder structure
- Add error handling and toast notifications for download operations
